plugins/cpp_reparse/service/include/service/cppreparseservice.h

```diff
@@ -0,0 +1,63 @@
+#ifndef CC_SERVICE_CPPREPARSESERVICE_H
+#define CC_SERVICE_CPPREPARSESERVICE_H
+
+#include <string>
+#include <memory>
+
+#include <boost/program_options/variables_map.hpp>
+
+#include <odb/database.hxx>
+
+#include <util/odbtransaction.h>
+#include <webserver/servercontext.h>
+
+#include <CppReparseService.h>
+
+namespace cc
+{
+
+namespace service
+{
+
+namespace reparse
+{
+
+class ASTCache;
+class CppReparser;
+
+} // namespace reparse
+
+namespace language
+{
+
+class CppReparseServiceHandler: virtual public CppReparseServiceIf
+{
+public:
+  CppReparseServiceHandler(
+    std::shared_ptr<odb::database> db_,
+    std::shared_ptr<std::string> datadir_,
+    const cc::webserver::ServerContext& context_);
+
+  ~CppReparseServiceHandler();
+
+  virtual bool isEnabled() override;
+
+  virtual void getAsHTML(
+    std::string& return_,
+    const core::FileId& fileId_) override;
+
+private:
+  std::shared_ptr<odb::database> _db;
+  util::OdbTransaction _transaction;
+  const boost::program_options::variables_map& _config;
+  std::shared_ptr<std::string> _datadir;
+
+  std::shared_ptr<reparse::ASTCache> _astCache;
+  std::unique_ptr<reparse::CppReparser> _reparser;
```

Comment:
Some architectural questions:
- Is it true that if the service is running on multiple threads then all of them is having an own cache and reparser? Isn't it worth sharing at least the cache among the threads?
- As far as I can see, only the reparser uses the cache. Then why does the `CppReparseServiceHandler` have to maintain it as a member?
- Does it have any reason these two members being pointers?

Reply:
1. No. The `XYZServiceHandler` classes are only instantiated **once** at the beginning of the server, I have specifically investigated this. Only one of this class lives in the server process no matter how many threads are there, and thus access to the internals of the class is shared -- which also makes it not thread-safe, thus locks must be put in certain places, such as in the cache. (This is also why the thread count is a sharp lower bound for the number of cache entries - to prevent constant cache flushes due to all threads doing a reparse on a different file.)
 2. It was designed with some further refactoring in mind as it is not exactly clear at this point (I think this will be introduced properly in the 4th reparse patch) how to architecturally lay out this plugin so other plugins can depend on it. The idea is that the AST Cache and the Reparser modules are "infrastructure" - other plugins (such as a "C++ Way-Too-Deep Analysis Visualisation Blackmagic Yadda" plugin) could depend on them, without depending on "visualisations and processes" made possible by this plugin, such as the HTML or the InfoTree.
 3. For two reasons: the ability to forward-declare these classes only, and to ensure that their lifetime is managed. Also, this later on (see *2.*) allows for sharing these things. Also, these members are **not** initialized (see ctor) if the user starts the server with `--disable-cpp-reparse=1`. This ensures that no memory (apart from a pointer's size and a bit of overhead) is wasted if reparse is turned off, as the current architecture doesn't allow turning plugins on and off at server start - thus, the `CppReparseServiceHandler` class is **always** instantiated if the server starts. (I think I'll put up an issue for a plugin enable-disable mechanis just now.)